fem/bilininteg_convection_pa.cpp

```diff
@@ -766,6 +774,113 @@ void SmemPAConvectionApply3D(const int ne,
+
+   const auto b = Reshape(b_, Q1D, D1D);
+   const auto X = Reshape(x_, D1D, D1D, VDIM, NE);
+   auto C = Reshape(y_, VDIM, Q1D, Q1D, NE);
+
+   MFEM_FORALL_2D(e, NE, Q1D, Q1D, 1,
+   {
+      constexpr int NBZ = 1;
+      constexpr int MQ1 = T_Q1D ? T_Q1D : T_MAX;
+      constexpr int MD1 = T_D1D ? T_D1D : T_MAX;
+
+      const int D1D = T_D1D ? T_D1D : d1d;
+      const int Q1D = T_Q1D ? T_Q1D : q1d;
+      const int VDIM = T_VDIM ? T_VDIM : vdim;
+
+      MFEM_SHARED double B[MQ1*MD1];
+      mfem::kernels::LoadB<MD1,MQ1>(D1D,Q1D,b,B);
+
+      MFEM_SHARED double DD[NBZ][MD1*MD1];
+      MFEM_SHARED double DQ[NBZ][MD1*MQ1];
+      MFEM_SHARED double QQ[NBZ][MQ1*MQ1];
+
+      for (int c = 0; c < VDIM; c++)
+      {
+         mfem::kernels::LoadX<MD1,NBZ>(e,D1D,c,X,DD);
+         mfem::kernels::EvalX<MD1,MQ1,NBZ>(D1D,Q1D,B,DD,DQ);
+         mfem::kernels::EvalY<MD1,MQ1,NBZ>(D1D,Q1D,B,DQ,QQ);
+
+         MFEM_FOREACH_THREAD(qx,x,Q1D)
+         {
+            MFEM_FOREACH_THREAD(qy,y,Q1D)
+            {
+               C(c,qx,qy,e) = kernels::PullEval<MQ1,NBZ>(qx,qy,QQ);
+            }
+         }
+         MFEM_SYNC_THREAD;
+      }
+   });
+}
+
+template<int T_VDIM = 0, int T_D1D = 0, int T_Q1D = 0, int T_MAX = 0>
+static void QEvalVGF3D(const int NE,
+                       const double *b_,
+                       const double *x_,
+                       double *y_,
+                       const int vdim = 1,
+                       const int d1d = 0,
+                       const int q1d = 0)
+{
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+   const int VDIM = T_VDIM ? T_VDIM : vdim;
+
+   const auto b = Reshape(b_, Q1D, D1D);
+   const auto X = Reshape(x_, D1D, D1D, D1D, VDIM, NE);
+   auto C = Reshape(y_, VDIM, Q1D, Q1D, Q1D, NE);
+
+   MFEM_FORALL_3D(e, NE, Q1D, Q1D, Q1D,
+   {
+      const int D1D = T_D1D ? T_D1D : d1d;
+      const int Q1D = T_Q1D ? T_Q1D : q1d;
+      const int VDIM = T_VDIM ? T_VDIM : vdim;
+      constexpr int MQ1 = T_Q1D ? T_Q1D : T_MAX;
+      constexpr int MD1 = T_D1D ? T_D1D : T_MAX;
+
+      MFEM_SHARED double B[MQ1*MD1];
+      mfem::kernels::LoadB<MD1,MQ1>(D1D,Q1D,b,B);
+
+      MFEM_SHARED double DDD[MD1*MD1*MD1];
+      MFEM_SHARED double DDQ[MD1*MD1*MQ1];
+      MFEM_SHARED double DQQ[MD1*MQ1*MQ1];
+      MFEM_SHARED double QQQ[MQ1*MQ1*MQ1];
+
+      for (int c = 0; c < VDIM; c++)
+      {
+         mfem::kernels::LoadX<MD1>(e,D1D,c,X,DDD);
+         mfem::kernels::EvalX<MD1,MQ1>(D1D,Q1D,B,DDD,DDQ);
+         mfem::kernels::EvalY<MD1,MQ1>(D1D,Q1D,B,DDQ,DQQ);
+         mfem::kernels::EvalZ<MD1,MQ1>(D1D,Q1D,B,DQQ,QQQ);
+
+         MFEM_FOREACH_THREAD(qx,x,Q1D)
+         {
+            MFEM_FOREACH_THREAD(qy,y,Q1D)
+            {
+               MFEM_FOREACH_THREAD(qz,z,Q1D)
+               {
+                  C(c,qx,qy,qz,e) = mfem::kernels::PullEval<MQ1>(qx,qy,qz,QQQ);
+               }
+            }
+         }
+         MFEM_SYNC_THREAD;
+      }
+   });
+}
```

Comment:
Shouldn't these be in `QuadInterpolator`?

Reply:
I second the proposal.

Reply:
I second the proposal.

Reply:
Waiting for #1737, then I'll push the code over there.

Reply:
`done`